File: project/frontend-services/apps/sharespot-device-records-frontend/src/app/remote-entry/mappers/RecordMapper.ts
import {DeviceRecordDTO, RecordTypeDTO, SensorDataRecordLabelDTO} from "../dtos/RecordsDTO";
import {DeviceRecord} from "../model/DeviceRecord";
import {RecordEntry} from "../model/RecordEntry";
import {RecordEntryType} from "../model/RecordEntryType";
import {SensorDataRecordLabel} from "../model/SensorDataRecordLabel";
import {Device} from "../model/Device";

export class RecordMapper {

  static dtoToModel(dto: DeviceRecordDTO): DeviceRecord {
    const entries = dto.entries.map(e => {
      if (e.type == RecordTypeDTO.BASIC) {
        return new RecordEntry(e.label, e.content, RecordEntryType.BASIC)
      } else {
        if ((e.label as SensorDataRecordLabelDTO) === SensorDataRecordLabelDTO.GPS_LONGITUDE) {
          return new RecordEntry(SensorDataRecordLabel.GPS_LONGITUDE, e.content, RecordEntryType.SENSOR_DATA)
        } else {
          return new RecordEntry(SensorDataRecordLabel.GPS_LATITUDE, e.content, RecordEntryType.SENSOR_DATA)
        }
      }
    });
    const device = new Device(dto.device.id, dto.device.name);
    return new DeviceRecord(device, entries);
  }

  static modelToDto(model: DeviceRecord): DeviceRecordDTO {
    const entries = model.entries.map(e => {
      if (e.type == RecordEntryType.BASIC) {
        return {label: e.label, content: e.content, type: RecordTypeDTO.BASIC}
      } else {
        if ((e.label as SensorDataRecordLabel) === SensorDataRecordLabel.GPS_LONGITUDE) {
          return {label: SensorDataRecordLabelDTO.GPS_LONGITUDE, content: e.content, type: RecordTypeDTO.SENSOR_DATA}
        } else {
          return {label: SensorDataRecordLabelDTO.GPS_LATITUDE, content: e.content, type: RecordTypeDTO.SENSOR_DATA}
        }
      }
    })
    return {device: {id: model.device.id, name: model.device.name}, entries: entries}
  }
}

File: project/backend-services/sharespot-device-records-slave-backend/infrastructure/persistence/postgres/src/main/java/sharespot/services/devicerecordsbackend/infrastructure/persistence/postgres/mapper/RecordMapper.java
package sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.mapper;

import sharespot.services.devicerecordsbackend.domain.model.Device;
import sharespot.services.devicerecordsbackend.domain.model.records.*;
import sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.model.DeviceRecordEntryPostgres;
import sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.model.DeviceRecordEntryTypePostgres;
import sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.model.DeviceRecordsPostgres;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

public class RecordMapper {

    public static DeviceRecords postgresToDomain(DeviceRecordsPostgres records) {
        List<RecordEntry> collect = records.entries.stream().map(e -> {
            if (e.type.equals(DeviceRecordEntryTypePostgres.sensorData())) {
                var label = SensorDataRecordLabel.give(e.label);
                return new SensorDataRecordEntry(label, e.content);
            } else {
                return new BasicRecordEntry(e.label, e.content);
            }
        }).collect(Collectors.toList());

        var deviceId = new DeviceId(UUID.fromString(records.deviceId));
        var deviceName = new DeviceName(records.name);

        return new DeviceRecords(new Device(deviceId, deviceName), new Records(collect));
    }

    public static DeviceRecordsPostgres domainToPostgres(DeviceRecords domain) {
        var deviceRecordsPostgres = new DeviceRecordsPostgres();
        deviceRecordsPostgres.name = domain.device().name().value();
        deviceRecordsPostgres.deviceId = domain.device().id().value().toString();
        deviceRecordsPostgres.entries = domain.records().entries().stream().map(e -> {
            var entry = new DeviceRecordEntryPostgres();
            if (e instanceof BasicRecordEntry) {
                entry.type = DeviceRecordEntryTypePostgres.basic();
            } else {
                entry.type = DeviceRecordEntryTypePostgres.sensorData();
            }
            entry.content = e.getContent();
            entry.label = e.getLabel();
            return entry;
        }).collect(Collectors.toSet());
        return deviceRecordsPostgres;
    }
}
